cpp/src/sampling/detail/graph_functions.hpp

```diff
@@ -0,0 +1,176 @@
+
+/**
+ * @brief Compute local out degrees of the majors belonging to the adjacency matrices
+ * stored on each gpu
+ *
+ * Iterate through partitions and store their local degrees
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return A single vector containing the local out degrees of the majors belong to the adjacency
+ * matrices
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> compute_local_major_degrees(
+  raft::handle_t const& handle, GraphViewType const& graph_view);
+
+/**
+ * @brief Calculate global degree information for all vertices represented by current gpu
+ *
+ * Calculate local degree and perform row wise exclusive scan over all gpus in column
+ * communicator.
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return Tuple of two device vectors. The first one contains per source edge-count encountered
+ * by gpus in the column communicator before current gpu. The second device vector contains the
+ * global out degree for every source represented by current gpu
+ */
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::edge_type>,
+           rmm::device_uvector<typename GraphViewType::edge_type>>
+get_global_degree_information(raft::handle_t const& handle, GraphViewType const& graph_view);
+
+/**
+ * @brief Calculate global adjacency offset for all majors represented by current gpu
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param[in] global_degree_offsets Global degree offset to local adjacency list for every major
+ * represented by current gpu
+ * @param global_out_degrees Global out degrees for every source represented by current gpu
+ * @return Device vector containing the number of edges that are prior to the adjacency list of
+ * every major that can be represented by the current gpu
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> get_global_adjacency_offset(
+  raft::handle_t const& handle,
+  GraphViewType const& graph_view,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_degree_offsets,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_out_degrees);
+
+/**
+ * @brief Gather active majors across gpus in a column communicator
```

Comment:
Is this gather or allgather (the results will be stored only in root or every process in the column communicator?)? If allgather, better rename to avoid confusion.

Reply:
Fixed in next push